### PR TITLE
buildenv: perl -> c++

### PR DESCRIPTION
**Pull request by shlevy**

### CODE REVIEW

File: src/buildenv/buildenv.cc [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+            continue;
+        const auto & srcFile = srcDir + "/" + ent.name;
+        auto dstFile = dstDir + "/" + ent.name;
+
+        /* The files below are special-cased to that they don't show up
+         * in user profiles, either because they are useless, or
+         * because they would cauase pointless collisions (e.g., each
+         * Python package brings its own
+         * `$out/lib/pythonX.Y/site-packages/easy-install.pth'.)
+         */
```
Comment (grahamc) [inline src/buildenv/buildenv.cc]: Not to nit, but is this comment accurate? I don't see how `$out/lib/pythonX.Y/site-packages/easy-install.pth` is excluded by the below suffix list.

Comment (shlevy) [inline src/buildenv/buildenv.cc]: Oops, I just copied the comment from the perl script but you're right. @edolstra any idea about this? Should that be on the list?

File: src/buildenv/buildenv.cc [hunk 2]
```diff
@@ -0,0 +1,188 @@
+#include "shared.hh"
+extern "C" {
```
Comment (edolstra) [inline src/buildenv/buildenv.cc]: I don't think this is necessary...

**Comment by edolstra** [comment]
I get:
```
$ ./inst/bin/nix-env -u subversion
building path(s) ‘/nix/store/xzglpy4plm5cpfx2iparghg9mj1q3cay-user-environment’
error: creating symlink from ‘/nix/store/xzglpy4plm5cpfx2iparghg9mj1q3cay-user-environment/nix-support’ to ‘/nix/store/hhs78ipwx5rz1692xns71086viy0yf6i-unrar-5.3.9/nix-support’: File exists
builder for ‘/nix/store/59p1rpz6x8b84kn1zib8v00dff2aw8lh-user-environment.drv’ failed with exit code 1
error: build of ‘/nix/store/59p1rpz6x8b84kn1zib8v00dff2aw8lh-user-environment.drv’ failed
```
Looks like `/nix-support` is not filtered correctly.

**Comment by edolstra** [comment]
BTW, it would be nice to change `buildenv` into a builtin derivation (see `builtinFetchurl`). That way, we wouldn't need any special hacks anymore for running in a chroot (the `chrootDeps` in `corepkgs/config.nix.in`).

**Comment by shlevy** [comment]
Oops, broke the filtering in a pre-push cleanup, fixing.

**Comment by shlevy** [comment]
@edolstra should be pretty straightforward to do the builtin derivation on top of this change, right? I'd rather focus my nix efforts on finishing up the perl rewrite.

**Comment by edolstra** [comment]
@shlevy Sure.

**Comment by shlevy** [comment]
@edolstra fixed

**Pull request merged**